web/src/input_pill.ts

```diff
@@ -5,12 +5,74 @@ import $ from "jquery";
 import render_input_pill from "../templates/input_pill.hbs";
 
 import * as blueslip from "./blueslip";
+import type {EmojiRenderingDetails} from "./emoji";
 import * as keydown_util from "./keydown_util";
 import * as ui_util from "./ui_util";
 
 // See https://zulip.readthedocs.io/en/latest/subsystems/input-pills.html
 
-export function create(opts) {
+export type InputPillItem<T> = {
+    display_value: string;
+    type: string;
+    img_src?: string;
+    deactivated?: boolean;
+    status_emoji_info?: EmojiRenderingDetails & {emoji_alt_code: boolean}; // TODO: Move this in user_status.js
+} & T;
+
+type InputPillCreateOptions<T> = {
+    $container: JQuery;
+    pill_config?: {
+        show_user_status_emoji?: boolean;
+    };
+    create_item_from_text: (
+        text: string,
+        existing_items: InputPillItem<T>[],
+    ) => InputPillItem<T> | undefined;
+    get_text_from_item: (item: InputPillItem<T>) => string;
+};
+
+type InputPill<T> = {
+    item: InputPillItem<T>;
+    $element: JQuery;
+};
+
+type InputPillStore<T> = {
+    pills: InputPill<T>[];
+    pill_config: InputPillCreateOptions<T>["pill_config"];
+    $parent: JQuery;
+    $input: JQuery;
+    create_item_from_text: InputPillCreateOptions<T>["create_item_from_text"];
+    get_text_from_item: InputPillCreateOptions<T>["get_text_from_item"];
+    onPillCreate?: () => void;
+    removePillFunction?: (pill: InputPill<T>) => void;
+    createPillonPaste?: () => void;
+};
+
+type InputPillRenderingDeatails = {
+    display_value: string;
+    has_image: boolean;
+    img_src?: string;
+    deactivated?: boolean;
+    has_status?: boolean;
+    status_emoji_info?: EmojiRenderingDetails & {emoji_alt_code: boolean};
+};
+
+// These are the functions that are exposed to other modules.
+type InputPillUserPrototype<T> = {
```

Comment:
Why is this called `InputPillUserPrototype`? If this is what is exposed to other modules, I would hope it to be something that would feel understandable when reading another module.